# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [301]:
# Import your libraries:
import pandas as pd
import numpy as np
import sklearn.datasets
from sklearn import datasets
from sklearn.metrics import mean_squared_error, r2_score


# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [302]:
# Your code here:
#sklearn.datasets.load_diabetes(return_X_y=False)
#diabetes_X, diabetes_y = datasets.load_diabetes(return_X_y=True)
diabetes = datasets.load_diabetes()

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [303]:
# Your code here:
diabetes.keys()

dict_keys(['data', 'target', 'DESCR', 'feature_names', 'data_filename', 'target_filename'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [304]:
# Your code here:
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - Age
      - Sex
      - Body mass index
      - Average blood pressure
      - S1
      - S2
      - S3
      - S4
      - S5
      - S6

Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times `n_samples` (i.e. the sum of squares of each column totals 1).

Source URL:
https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html

For more information see:
Bra

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

# Enter your answer here:
1. Attributes number are 10  that are numeric predictive values :
      - Age
      - Sex
      - Body mass index
      - Average blood pressure
      - S1
      - S2
      - S3
      - S4
      - S5
      - S6
2. From 'data' the 'target' is obtained, that is, 'data' are the 
independent variables and 'target' are the dependent variables.
3. There are 442 records.

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [305]:
# Your code here:

y = pd.DataFrame(diabetes.target)
X = pd.DataFrame(diabetes.data)

In [306]:
y.describe().T

count        mean        std   min   25%    50%    75%    max
0  442.0  152.133484  77.093005  25.0  87.0  140.5  211.5  346.0

In [307]:
y.dtypes

0    float64
dtype: object

In [308]:
X.describe().T

count          mean       std       min       25%       50%       75%  \
0  442.0 -3.639623e-16  0.047619 -0.107226 -0.037299  0.005383  0.038076   
1  442.0  1.309912e-16  0.047619 -0.044642 -0.044642 -0.044642  0.050680   
2  442.0 -8.013951e-16  0.047619 -0.090275 -0.034229 -0.007284  0.031248   
3  442.0  1.289818e-16  0.047619 -0.112400 -0.036656 -0.005671  0.035644   
4  442.0 -9.042540e-17  0.047619 -0.126781 -0.034248 -0.004321  0.028358   
5  442.0  1.301121e-16  0.047619 -0.115613 -0.030358 -0.003819  0.029844   
6  442.0 -4.563971e-16  0.047619 -0.102307 -0.035117 -0.006584  0.029312   
7  442.0  3.863174e-16  0.047619 -0.076395 -0.039493 -0.002592  0.034309   
8  442.0 -3.848103e-16  0.047619 -0.126097 -0.033249 -0.001948  0.032433   
9  442.0 -3.398488e-16  0.047619 -0.137767 -0.033179 -0.001078  0.027917   

        max  
0  0.110727  
1  0.050680  
2  0.170555  
3  0.132044  
4  0.153914  
5  0.198788  
6  0.181179  
7  0.185234  
8  0.133599  
9  0.135612

In [309]:
X.dtypes

0    float64
1    float64
2    float64
3    float64
4    float64
5    float64
6    float64
7    float64
8    float64
9    float64
dtype: object

In [310]:
y.isnull().sum()

0    0
dtype: int64

In [311]:
X.isnull().sum()

0    0
1    0
2    0
3    0
4    0
5    0
6    0
7    0
8    0
9    0
dtype: int64

# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:

```
y = β0 + β1X1 + β2X2 + ... + βnXn + ϵ
```

...where:

- X1-Xn: data 
- β0: intercept 
- β1-βn: coefficients 
- ϵ: error (cannot explained by model)
- y: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [312]:
# Your code here:
from sklearn import linear_model

#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [313]:
# Your code here:
regr = linear_model.LinearRegression()

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [314]:
# Your code here:
diabetes_data_train = diabetes_X[:-20]
diabetes_data_test = diabetes_X[-20:]
diabetes_target_train = diabetes_y[:-20]
diabetes_target_test = diabetes_y[-20:]

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [315]:
# Your code here:
regr.fit(diabetes_data_train, diabetes_target_train)

print('Intercept: \n', regr.intercept_)
print('Coefficients: \n', regr.coef_)


Intercept: 
 152.76430691633442
Coefficients: 
 [ 3.03499549e-01 -2.37639315e+02  5.10530605e+02  3.27736980e+02
 -8.14131709e+02  4.92814588e+02  1.02848452e+02  1.84606489e+02
  7.43519617e+02  7.60951722e+01]


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [316]:
# Your code here:
diabetes_y_pred = regr.predict(diabetes_data_test)
diabetes_y_pred

array([197.61846908, 155.43979328, 172.88665147, 111.53537279,
       164.80054784, 131.06954875, 259.12237761, 100.47935157,
       117.0601052 , 124.30503555, 218.36632793,  61.19831284,
       132.25046751, 120.3332925 ,  52.54458691, 194.03798088,
       102.57139702, 123.56604987, 211.0346317 ,  52.60335674])

#### Print your `diabetes_target_test` and compare with the prediction. 

In [317]:
# Your code here:
print(diabetes_target_test)

[233.  91. 111. 152. 120.  67. 310.  94. 183.  66. 173.  72.  49.  64.
  48. 178. 104. 132. 220.  57.]


#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

# Your explanation here:
#### It is not the same because a mistake is made when making the prediction.


# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://onlinecourses.science.psu.edu/stat501/node/297/) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    ```
    H0 : β1 = β2 = ... = β10 = 0
    HA : At least one βj ≠ 0 (for j = 1, 2, ..., 10)
    ```

* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

![statsmodels regression](../statsmodels.png)

In [318]:
# Your code here:
!pip install statsmodels


In [319]:
import statsmodels.api as sm

In [320]:
# X: usualmente significa nuestras variables de entrada (o variables independientes)
X = sm.add_constant(diabetes_data_train)            # Se agrega una intercepción (beta_0) al modelo
y = diabetes_target_train                           # Y: normalmente significa la variable respuesta (o variable dependiente)

# Para encajar en el modelo

mod = sm.OLS(y, X).fit()              # sm.OLS(respuesta, entrada)
predictions = mod.predict(X)
print(mod1.summary())


                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.512
Model:                            OLS   Adj. R-squared:                  0.500
Method:                 Least Squares   F-statistic:                     43.16
Date:                Fri, 17 Apr 2020   Prob (F-statistic):           4.64e-58
Time:                        15:56:08   Log-Likelihood:                -2281.1
No. Observations:                 422   AIC:                             4584.
Df Residuals:                     411   BIC:                             4629.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        152.7643      2.658     57.469      0.0

#### Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?

1. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?

1. How will you modify your linear reguression model according to the test results above?

# Your answers here:
### 1. The F-score is 43.16 and the probability is approximately zero, so the null hypothesis is rejected
### 2. It is zero for coefficients that do not add much to the model??
### 3. It would be better to remove them from the model??

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [321]:
# Your code here:
auto = pd.read_csv('../auto-mpg.csv')

Look at the first 5 rows using the `head()` function:

In [322]:
# Your code here:
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [323]:
# Your code here:
auto.dtypes

mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

In [324]:
auto.describe().T

count         mean         std     min       25%     50%  \
mpg           398.0    23.514573    7.815984     9.0    17.500    23.0   
cylinders     398.0     5.454774    1.701004     3.0     4.000     4.0   
displacement  398.0   193.425879  104.269838    68.0   104.250   148.5   
horse_power   392.0   104.469388   38.491160    46.0    75.000    93.5   
weight        398.0  2970.424623  846.841774  1613.0  2223.750  2803.5   
acceleration  398.0    15.568090    2.757689     8.0    13.825    15.5   
model_year    398.0    76.010050    3.697627    70.0    73.000    76.0   

                   75%     max  
mpg             29.000    46.6  
cylinders        8.000     8.0  
displacement   262.000   455.0  
horse_power    126.000   230.0  
weight        3608.000  5140.0  
acceleration    17.175    24.8  
model_year      79.000    82.0

What is the newest model year and the oldest model year?

In [325]:
# Your code here:
print('The newest model is from de year: \n',auto.model_year.max())
print('The oldest model is from de year: \n',auto.model_year.min())

The newest model is from de year: 
 82
The oldest model is from de year: 
 70


Check the dataset for missing values and remove all rows containing at least one missing value.

In [326]:
# Your code here:
auto.isnull().sum()


mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

In [327]:
auto.dropna(inplace = True)

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [328]:
# Your code here:
auto.cylinders.value_counts()

4    199
8    103
6     83
3      4
5      3
Name: cylinders, dtype: int64

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [329]:
# Your code here:
auto.drop('car_name', axis = 1, inplace = True)
auto

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0    18.0          8         307.0        130.0    3504          12.0   
1    15.0          8         350.0        165.0    3693          11.5   
2    18.0          8         318.0        150.0    3436          11.0   
3    16.0          8         304.0        150.0    3433          12.0   
4    17.0          8         302.0        140.0    3449          10.5   
..    ...        ...           ...          ...     ...           ...   
393  27.0          4         140.0         86.0    2790          15.6   
394  44.0          4          97.0         52.0    2130          24.6   
395  32.0          4         135.0         84.0    2295          11.6   
396  28.0          4         120.0         79.0    2625          18.6   
397  31.0          4         119.0         82.0    2720          19.4   

     model_year  
0            70  
1            70  
2            70  
3            70  
4            70  
..          ...  
393          82  
394          82  
395          82  
396          82  
397          82  

[392 rows x 7 columns]

In [330]:
# Your code here:
auto.columns.tolist()

['mpg',
 'cylinders',
 'displacement',
 'horse_power',
 'weight',
 'acceleration',
 'model_year']

In [331]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

In [332]:

FEATS = ['cylinders','displacement','horse_power','weight','acceleration','model_year']
TARGET = 'mpg'
X_train, X_test, y_train, y_test = train_test_split(auto[FEATS], auto[TARGET], test_size=0.2)

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [333]:
# Your code here:
auto_model = LinearRegression()
auto_model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](). Its formula is:

![R Squared](../r-squared.png)

...where:

* yi is an actual data point.
* ŷi is the corresponding data point on the estimated regression line.

By adding the squares of the difference between all yi-ŷi pairs, we have a measure called SSE (*error sum of squares*) which is an application of the r squared score to indicate the extent to which the estimated regression model is different from the actual data. And we attribute that difference to the random error that is unavoidable in the real world. Obviously, we want the SSE value to be as small as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [334]:
# Your code here:
y_pred = auto_model.predict(X_train)


In [335]:

from sklearn.metrics import r2_score

In [337]:
print(f"train error: {r2_score(y_pred=y_pred, y_true=y_train)}")

train error: 0.8174274793874765


#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [340]:
# Your code here:
y_test_pred = auto_model.predict(X_test)
print(f"test error: {r2_score(y_pred=y_test_pred, y_true=y_test)}")

test error: 0.7671430509017924


#### Explaining the results

The r squared scores of the training data and the test data are pretty close (0.8146 vs 0.7818). This means our model is not overfitted. However, there is still room to improve the model fit. Move on to the next challenge.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [341]:
# Your code here:
X_train09, X_test09, y_train09, y_test09 = train_test_split(auto[FEATS], auto[TARGET], test_size=0.1)

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [343]:
# Your code here:
auto_model09 = LinearRegression()
auto_model09.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

Compute the predicted values and r squared score for our new model and new sample data.

In [345]:
# Your code here:
y_pred09 = auto_model09.predict(X_train09)

print(f"train error: {r2_score(y_pred=y_pred09, y_true=y_train09)}")

train error: 0.8082766298342371


Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [348]:
# Your code here:
y_test_pred09 = auto_model09.predict(X_test09)
print(f"test error: {r2_score(y_pred=y_test_pred09, y_true=y_test09)}")
print('Yes, the test error has improved but not the train error')

test error: 0.7940205145532567
Yes, the test error has improved but not the train error


# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

In [349]:
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [353]:
# Your code here:
selector = RFE(auto_model,  n_features_to_select = 3)

Fit the model and print the ranking

In [356]:
# Your code here:
selector.fit(X_train, y_train)
selector.ranking_

array([1, 4, 3, 2, 1, 1])

Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [360]:
# Your code here:
FEATS_REDUCED =  ['cylinders','acceleration','model_year']
X_train_reduced, X_test_reduced, y_train_reduced, y_test_reduced = train_test_split(auto[FEATS], 
                                                                                    auto[TARGET], test_size=0.2)

Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [362]:
# Your code here: 

auto_model_reduced = LinearRegression()
auto_model_reduced.fit(X_train_reduced, y_train_reduced)
y_pred_reduced = auto_model09.predict(X_train_reduced)
y_test_pred_reduced = auto_model09.predict(X_test_reduced)




In [365]:
print(f"train error: {r2_score(y_pred=y_pred_reduced, y_true=y_train_reduced)}")
print(f"test error: {r2_score(y_pred=y_test_pred_reduced, y_true=y_test_reduced)}")
print('The error has not improved')

train error: 0.8094850300825259
test error: 0.793670918203457
The error has not improved


#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.